### DISCUSSION TITLE
Can I access the model output_path in a custom logger?

### DISCUSSION THREAD

**Discussion by einarbmag** [explosion/spaCy#7414] [Help: Coding & Implementations] [answered]
I'm trying to write a custom logger for spaCy+MLflow integration, as the integration provided by MLflow is not so useful. I want to log the parameters and scores obviously, but I also want to log the best model at the end of the training. I guess this should logically happen in `finalize_logger`, but I don't know if there's any way for me to know the value of output_path there. I would really prefer not to modify the training loop as my aim is to contain all my custom logic in registered functions so I can just use spaCy config files and the spacy train command.

**Comment by einarbmag** [comment]
@adrianeboyd 
Any ideas on this? Would be great to be able to customize how the model is persisted without modifying the training loop.

**Comment by svlandeg** [comment, marked as answer]
Hm, I think I understand what you mean, and why you'd want that. Currently I'm afraid the loggers aren't set up to do this though, they're only logging information for each step, and they don't have access to the model-saving logic. So I don't see a way around modifying the training loop, if you'd need this...

**Comment by narayanacharya6** [comment]
I came across this requirement in one of my projects. I was wondering if we can do something similar to the wandb logger (under registry name `spacy.WandbLogger.v2`) as a custom logger for mlflow which takes in the `output_path` as an input parameter. Is this doable or am I missing something here?

**Comment by lluissalord** [comment]
Do you have any update or implementation regarding this as I think it could be useful for our project.

**Comment by svlandeg** [comment]
Hm I think I want to change my original answer. If you follow the current logger API, `log_step(info: Optional[Dict[str, Any]])` is called on each iteration. You can access `info.get("output_path")` and store the path for use when `finalize_logger` is being called - though you'd need to substitute `DIR_MODEL_LAST` for `DIR_MODEL_BEST` (cf https://github.com/explosion/spaCy/blob/master/spacy/training/loop.py).

That's little bit of a hack, but should work.

Alternatively, we could pass onwards an `info` dict also for the `finalize` step of the logger, but then that's going to break older versions of user-implemented loggers :(
